flight/PiOS/STM32F4xx/pios_dmashot.c

```diff
@@ -0,0 +1,591 @@
+ */
+void PIOS_DMAShot_InitializeGPIOs()
+{
+	// If there's nothing setup, fail hard. We shouldn't be getting here.
+	PIOS_Assert(dmashot_cfg && servo_timers);
+
+	for (int i = 0; i < MAX_TIMERS; i++) {
+		for (int j = 0; j < 4; j++) {
+
+			struct servo_timer *s_timer = servo_timers[i];
+			if (!s_timer || !s_timer->sysclock)
+				continue;
+
+			const struct pios_tim_channel *servo_channel = s_timer->servo_channels[j];
+			if (servo_channel) {
+
+				GPIO_InitTypeDef gpio_cfg = servo_channel->pin.init;
+				gpio_cfg.GPIO_Speed = GPIO_High_Speed;
+				GPIO_Init(servo_channel->pin.gpio, &gpio_cfg);
+
+				GPIO_PinAFConfig(servo_channel->pin.gpio, servo_channel->pin.pin_source, servo_channel->remap);
+			}
+		}
+	}
+}
+
+static void PIOS_DMAShot_TimerSetup(struct servo_timer *s_timer, uint32_t sysclock, uint32_t dshot_freq, TIM_OCInitTypeDef *ocinit, bool master)
+{
+	TIM_TypeDef *timer = master ? s_timer->dma->master_timer : s_timer->dma->timer;
+	TIM_TimeBaseInitTypeDef timerdef;
+
+	TIM_Cmd(timer, DISABLE);
+
+	TIM_TimeBaseStructInit(&timerdef);
+
+	timerdef.TIM_Prescaler = 0;
+	timerdef.TIM_Period = sysclock / dshot_freq;
+	timerdef.TIM_ClockDivision = TIM_CKD_DIV1;
+	timerdef.TIM_RepetitionCounter = 0;
+	timerdef.TIM_CounterMode = TIM_CounterMode_Up;
+
+	TIM_TimeBaseInit(timer, &timerdef);
+
+	// TIM_Channels are spread apart by 4 in stm32f4xx_tim.h
+	for(int i = s_timer->low_channel; i <= s_timer->high_channel; i+=4)
+	{
+		switch(i)
+		{
+			case TIM_Channel_1:
+				TIM_OC1Init(timer, ocinit);
+				TIM_OC1PreloadConfig(timer, TIM_OCPreload_Enable);
+				break;
+			case TIM_Channel_2:
+				TIM_OC2Init(timer, ocinit);
+				TIM_OC2PreloadConfig(timer, TIM_OCPreload_Enable);
+				break;
+			case TIM_Channel_3:
+				TIM_OC3Init(timer, ocinit);
+				TIM_OC3PreloadConfig(timer, TIM_OCPreload_Enable);
+				break;
+			case TIM_Channel_4:
+				TIM_OC4Init(timer, ocinit);
+				TIM_OC4PreloadConfig(timer, TIM_OCPreload_Enable);
+				break;
+		}
+	}
+
+	// Do this, in case SyncPWM was configured before.
+	TIM_SelectOnePulseMode(timer, TIM_OPMode_Repetitive);
+}
+
+/**
+ * @brief Initializes and configures the registered timers for DMAShot operation.
+ */
+void PIOS_DMAShot_InitializeTimers(TIM_OCInitTypeDef *ocinit)
+{
+	// If there's nothing setup, fail hard. We shouldn't be getting here.
+	PIOS_Assert(dmashot_cfg && servo_timers);
+
+	for (int i = 0; i < MAX_TIMERS; i++) {
+
+		struct servo_timer *s_timer = servo_timers[i];
+		if (!s_timer || !s_timer->sysclock)
+			continue;
+
+		if(PIOS_DMAShot_GetNumChannels(s_timer) == 0) {
+			// Servo's should have been registered before InitializeTimers. If there's none,
+			// disable the timer.
+			s_timer->sysclock = 0;
+			continue;
+		}
+
+		PIOS_DMAShot_TimerSetup(s_timer, s_timer->sysclock, s_timer->dshot_freq, ocinit, false);
+
+		int f = s_timer->sysclock / s_timer->dshot_freq;
+
+		s_timer->duty_cycle_0 = f * DSHOT_DUTY_CYCLE_0 / 100;
```

Comment:
(f * DSHOT_DUTY_CYCLE_0 + 50) / 100

Basically, get correct rounding when you can't hit timing exactly.